applications/GeoMechanicsApplication/custom_utilities/dof_utilities.cpp

```diff
@@ -0,0 +1,116 @@
+        // instance. However, we assume that each DOF corresponds to a `Variable<double>` instance.
+        // Then we should be able to downcast the `VariableData` instance.
+        auto p_variable = dynamic_cast<const Variable<double>*>(&p_dof->GetVariable());
+        KRATOS_ERROR_IF_NOT(p_variable) << "Variable associated with DOF is not of type double" << std::endl;
+        return p_dof->GetSolutionStepValue(p_variable->GetTimeDerivative().GetTimeDerivative(), Step);
+    };
+    std::transform(rDofs.begin(), rDofs.end(), result.begin(), get_second_time_derivative);
+    return result;
+}
+
+Vector ExtractFirstTimeDerivativesOfUPwDofs(const std::vector<Dof<double>*>& rDofs, int Step)
+{
+    auto result                    = Vector(rDofs.size());
+    auto get_first_time_derivative = [Step](const auto p_dof) -> double {
+        // Why should we return 0.0 for the first time derivatives of water pressure?
+        if (p_dof->GetVariable() == WATER_PRESSURE) return 0.0;
+
+        // Unfortunately, the first time derivative cannot be accessed from a `VariableData`
+        // instance. However, we know that for all U-Pw elements the degrees of freedom are either
+        // displacement components or water pressures, which correspond to `Variable<double>`
+        // instances. Therefore, we should be able to downcast the `VariableData` instance.
+        auto p_variable = dynamic_cast<const Variable<double>*>(&p_dof->GetVariable());
+        KRATOS_ERROR_IF_NOT(p_variable) << "Variable associated with DOF is not of type double" << std::endl;
+        return p_dof->GetSolutionStepValue(p_variable->GetTimeDerivative(), Step);
+    };
+    std::transform(rDofs.begin(), rDofs.end(), result.begin(), get_first_time_derivative);
+    return result;
+}
+
+Vector ExtractSecondTimeDerivativesOfUPwDofs(const std::vector<Dof<double>*>& rDofs, int Step)
+{
+    auto result                     = Vector(rDofs.size());
+    auto get_second_time_derivative = [Step](const auto p_dof) -> double {
+        // Why should we return 0.0 for the second time derivatives of water pressure?
+        if (p_dof->GetVariable() == WATER_PRESSURE) return 0.0;
+
+        // Unfortunately, the first time derivative cannot be accessed from a `VariableData`
+        // instance. However, we know that for all U-Pw elements the degrees of freedom are either
+        // displacement components or water pressures, which correspond to `Variable<double>`
+        // instances. Therefore, we should be able to downcast the `VariableData` instance.
+        auto p_variable = dynamic_cast<const Variable<double>*>(&p_dof->GetVariable());
+        KRATOS_ERROR_IF_NOT(p_variable) << "Variable associated with DOF is not of type double" << std::endl;
+        return p_dof->GetSolutionStepValue(p_variable->GetTimeDerivative().GetTimeDerivative(), Step);
+    };
+    std::transform(rDofs.begin(), rDofs.end(), result.begin(), get_second_time_derivative);
+    return result;
+}
```

Comment:
On first glance these are four repetitions of a similar thing. The remark about VariableData I don't understand yet. As far as I have understood all D.O.F. are double by nature.

Reply:
We have discussed this in-person. The bottom line is that the functions ending with `OfUPwDofs` are only there to not change the current behavior. However, we do need to investigate why the water pressure (and its time derivatives) are assumed to be zero. When we can show that this is not needed at all (or if we can find a better alternative), then we can start changing the function's behavior. And, in fact, that may mean we no longer need these dedicated functions.